### ISSUE TITLE
Rule Request: useless function

### ISSUE THREAD

**Issue by KangOl** [astral-sh/ruff#12789]
Today, I (mostly) wrote the following code:

```python
class BaseConverter:
    def to_html(self):
        return HTMLConverter()

    def _loads(self, string):
        raise NotImplementedError

    def __call__(self):
        ...

def HTMLConverter(BaseConverter):
    def to_html(self):
        return self
    
    def _loads(self, string):
        return lxml.html.fromstring(string)
```

Did you spot the error? Obviously, the title of this issue is a big hint, but for the other ones, here is how to fix the code:
```diff
- def HTMLConverter(BaseConverter):
+ class HTMLConverter(BaseConverter):
```

I think this kind of error can be detected by ruff, as the body of the `HTMLConverter` *function* only contains other functions and its sole argument is the name of a defined class.

**Comment by charliermarsh** [comment]
That's funny. We could probably support functions that only contain definitions. What do you think, @AlexWaygood?

**Comment by AlexWaygood** [comment]
Yeah, I think this is a reasonable request. I agree that a function that only contains other functions is almost certainly not what was intended, so I think it's safe to flag it 😄

> We could probably support functions that only contain definitions.

Hmm, functions that only contain class definitions aren't quite so clear cut, since they can still have side effects when they're executed:

```pycon
>>> def foo():
...   class Bar:
...     print('yo')
...     
>>> foo()
yo
```

Though I suppose even for a function that only contains function definitions (and doesn't return any of them), it's hard to say _for sure_ that it didn't have any side effects, because of decorators:

```pycon
>>> REGISTRY = []
>>> def foo():
...   @REGISTRY.append
...   def bar(): pass
...     
>>> foo()
>>> REGISTRY
[<function foo.<locals>.bar at 0x11b6a3370>]
```

**Comment by dhruvmanila** [comment]
@AlexWaygood feel free to update the labels accordingly :)

**Comment by AlexWaygood** [comment]
I think exactly what the decision should be here depends on how many false positives we'd be willing to tolerate from this rule:
- If we want it to have a "pyflakes level of accuracy", where the rule has close to zero false positives, then we should _only_ emit a diagnostic if we see a function where the only items in the function body are function definitions, and none of the inner functions have any decorators
- If we're prepared to accept more false positives from the rule -- because we want it to catch more errors -- then we could also emit the diagnostic even if some of the inner functions are decorated, or if some of the inner definitions are class definitions.

So while I'm supportive of adding a rule like this, I'm not 100% sure what exactly the rule should look like. @charliermarsh / @dhruvmanila -- any thoughts from you on that?
